torch/csrc/jit/OVERVIEW.md

```diff
@@ -886,267 +884,244 @@ def LSTMCellS(x, hx, cx, w_ih, w_hh, b_ih, b_hh):
-
-* dtype
-* rank, but not size
-* requires_grad
-* device type (cpu, cuda)
-* defined - whether the Tensor exists or is a placeholder
+### Profiling ###
 
-The ArgumentSpec object is used as a key into a cache that holds pre-optimized Code objects (held in an ExecutionPlan object). On a cache hit, an InterpreterState is created and the Code in the cache is run.
-
-
-*Pre-derivative Optimization* On a code cache miss, we generate a new optimized Graph on the fly (`compileSpec`). It starts by creating a copy of the initial Graph and setting the input types to the specialized Tensor types observed in this specialization. TensorType inputs to the Graph will get refined with types that know the device, number of dimensions, and requires grad state.
+In the profiling stage, a Profiling Graph Executor takes an unoptimized graph runs required and passes that doesn't need profiling information ([see the next section](#Optimization)) and inserts `prim::profile` on **tensor-typed** inputs to fusible operations. 
+`torch::jit::ProfilingRecord` also maintains a mapping between output values of `prim::profile` nodes and the properties of input tensors at runtime.
+This mapping is maintained for each invocation of a function identified by Interpreter's `frame_id` (which is globally unique).
+It's important to not mix the tensor properties from different invocations for the same `torch::jit::Value*`, as this may inhibit symbolic (dynamic) shape inference to correctly inferring the relationships between the arguments of an operation and its outputs.                                                                                     
+Whenever a `prim::profile` node is executed, its `prim::profile`'s `Value*` and `frame_id` are used as a key and input tensor's properties are captured in a freshly created `TensorType` and stored in `profiled_types_per_frame_`.
+If the `prim::profile` node is executed more than once per execution (i.e. it's in a loop), the existing and incoming `TensorType` are merged.
+The merging logic follows a very simple rule: if two individual properties are different, the result is unknown (denoted by an empty `c10::optional<T>`) 
+At the very end of the execution of the profiled function, a countdown counter whose starting value is set to the number of profiled runs is decremented. When the counter hits 0, the profiler merges the profiled types from all the runs into a single profiled tensor type per a `Value*`. If the function is profiled more than once, some dimensions may contain a `torch::jit::ShapeSymbol` instead of a fixed number. For example, when a batch dimension changes from one run to another but stays the same within the same run across multiple operations, all participating tensor `Value`s should have the exact same `ShapeSymbol` in the batch dimension.
```

Comment:
Might be good to mentioned where and when this counter is created and where it resides.